addons/account/models/account_move.py

```diff
@@ -1,349 +1,1869 @@
+            totals_map['total_amount_currency'] += diff_amount_currency
+            lines_map['rounding_lines'] = rounding_lines
+
+            if in_draft_mode:
+                rounding_lines._onchange_amount_currency()
+                rounding_lines._onchange_balance()
+        else:
+            self.line_ids -= rounding_lines
+            lines_map['rounding_lines'] = self.env['account.move.line']
+
+    @api.multi
+    def _recompute_payment_terms_lines(self, lines_map, totals_map):
+        ''' Compute the dynamic payment term lines of the journal entry.
+
+        :param lines_map:               The line_ids dispatched by type containing:
+            * base_lines:               The lines having a tax_ids set.
+            * tax_lines:                The lines having a tax_line_id set.
+            * terms_lines:              The lines generated by the payment terms of the invoice.
+            * rounding_lines:           The cash rounding lines of the invoice.
+        :param totals_map: A dictionary having two keys:
+            * total_balance:            Sum(balance) of each product / tax line.
+            * total_amount_currency:    Sum(amount_currency) of each product / tax line.
+        '''
+        self.ensure_one()
+        in_draft_mode = self != self._origin
+
+        today = fields.Date.context_today(self)
+
+        available_term_lines_index = 0
+        available_term_lines = lines_map['terms_lines'].sorted(lambda line: line.date_maturity or today)
+
+        # Compute the date to be used during the computation of the payment terms.
+        if self.invoice_payment_term_id:
+            terms_date = self.invoice_date or fields.Date.context_today(self)
+        else:
+            terms_date = self.invoice_date_due or self.invoice_date or today
+
+        # Handle the special case in which there is no remaining base/tax lines. In such case, simply drop all
+        # payment term lines.
+        if not lines_map['base_lines'] and not lines_map['tax_lines']:
+            self.line_ids -= lines_map['terms_lines']
+            return
+
+        # Compute payment terms account.
+        # Find the receivable/payable account.
+        # /!\ The partner could not be set on a receipt.
+        if lines_map['terms_lines']:
+            # Retrieve account from previous payment terms lines in order to allow the user to set a custom one.
+            account = lines_map['terms_lines'][0].account_id
+        elif self.partner_id:
+            # Retrieve account from partner.
+            if self.type in ('out_invoice', 'out_refund', 'out_receipt'):
+                account = self.partner_id.property_account_receivable_id
+            elif self.type in ('in_invoice', 'in_refund', 'in_receipt'):
+                account = self.partner_id.property_account_payable_id
+            else:
+                account = None
+        elif self.type in ('out_receipt', 'in_receipt', 'out_invoice', 'out_refund', 'in_invoice', 'in_refund'):
```

Comment:
`self.type != 'entry'` as the list of values becomes quite long?

Reply:
actually, the use of `_is_invoice()` would be better

Reply:
I avoid to use that because if one day we had a new type, we don't want to fix all such conditions like that. However, I made a helper `move._is_invoice()`